util/rate_limiter.cc

```diff
@@ -365,8 +417,81 @@ RateLimiter* NewGenericRateLimiter(
   assert(rate_bytes_per_sec > 0);
   assert(refill_period_us > 0);
   assert(fairness > 0);
-  return new GenericRateLimiter(rate_bytes_per_sec, refill_period_us, fairness,
-                                mode, SystemClock::Default(), auto_tuned);
+  std::unique_ptr<RateLimiter> limiter(
+      new GenericRateLimiter(rate_bytes_per_sec, refill_period_us, fairness,
+                             mode, SystemClock::Default(), auto_tuned));
+  Status s = limiter->PrepareOptions(ConfigOptions());
+  if (s.ok()) {
+    return limiter.release();
+  } else {
+    assert(false);
+    return nullptr;
+  }
+}
+namespace {
+#ifndef ROCKSDB_LITE
+static int RegisterBuiltinRateLimiters(ObjectLibrary& library,
+                                       const std::string& /*arg*/) {
+  library.Register<RateLimiter>(
+      GenericRateLimiter::kClassName(),
+      [](const std::string& /*uri*/, std::unique_ptr<RateLimiter>* guard,
+         std::string* /*errmsg*/) {
+        guard->reset(new GenericRateLimiter(port::kMaxInt64));
+        return guard->get();
+      });
+  size_t num_types;
+  return static_cast<int>(library.GetFactoryCount(&num_types));
+}
+
+static std::unordered_map<std::string, RateLimiter::Mode>
+    rate_limiter_mode_map = {
+        {"kReadsOnly", RateLimiter::Mode::kReadsOnly},
+        {"kWritesOnly", RateLimiter::Mode::kWritesOnly},
+        {"kAllIo", RateLimiter::Mode::kAllIo},
+};
+#endif  // ROCKSDB_LITE
+static bool LoadRateLimiter(const std::string& name,
+                            std::shared_ptr<RateLimiter>* limiter) {
+  auto plen = strlen(GenericRateLimiter::kClassName());
+  if (name.size() > plen + 2 && name[plen] == ':' &&
+      StartsWith(name, GenericRateLimiter::kClassName())) {
+    auto rate = ParseInt64(name.substr(plen + 1));
+    limiter->reset(new GenericRateLimiter(rate));
+    return true;
+  } else {
+    return false;
+  }
+}
+
+static std::unordered_map<std::string, OptionTypeInfo> rate_limiter_type_info =
```

Comment:
I'm not sure, can we combine this on with the generic_rate_limiter_type_info? @hx235 what do you think?

Reply:
Catching up :)

Reply:
> With this PR change, is this rate_limiter_type_info required in implementation of every subclasses extending RateLimiter?

No need to reply to this - just found the answer by looking at your `MockRateLimiter` in test - thanks!